File: game-server/src/main/java/work/fking/pangya/game/player/CaddieRoster.java
package work.fking.pangya.game.player;

import work.fking.pangya.common.Rand;
import work.fking.pangya.game.model.IffContainer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static work.fking.pangya.game.model.IffObject.TYPE_CADDIE;
import static work.fking.pangya.game.model.IffObject.iffTypeFromId;

public class CaddieRoster implements IffContainer<Caddie> {

    private final List<Caddie> caddies = new ArrayList<>();

    @Override
    public List<Caddie> entries() {
        return Collections.unmodifiableList(caddies);
    }

    public void unlockCaddie(int iffId) {
        if (iffTypeFromId(iffId) != TYPE_CADDIE) {
            throw new IllegalArgumentException("iffId is not a caddie");
        }
        var caddie = new Caddie(Rand.max(10000), iffId, 0, 0);
        caddies.add(caddie);
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/player/CaddieRoster.kt
package work.fking.pangya.game.player

import work.fking.pangya.game.model.IFF_TYPE_CADDIE
import work.fking.pangya.game.model.IffContainer
import work.fking.pangya.game.model.iffTypeFromId
import java.util.concurrent.atomic.AtomicInteger

private val uidSequence = AtomicInteger(1)

class CaddieRoster(override val entries: MutableList<Caddie> = ArrayList()) : IffContainer<Caddie> {

    fun unlockCaddie(iffId: Int) {
        require(iffTypeFromId(iffId) == IFF_TYPE_CADDIE) { "iffId is not a caddie" }
        val caddie = Caddie(uid = uidSequence.getAndIncrement(), iffId = iffId, levelsGained = 4, experience = 1000)
        entries.add(caddie)
    }
}
